### PR TITLE
ARROW-14185: [C++] HashJoinNode should validate HashJoinNodeOptions

### PR DESCRIPTION
**Pull request by JabariBooker**
Adding validation for HashJoinNodeOptions

### CODE REVIEW

**Comment by lidavidm** [comment]
General comment: can we also add a few short unit tests to test this validation?

File: cpp/src/arrow/compute/exec/hash_join_node.cc [hunk 1]
```diff
@@ -453,6 +453,20 @@ Status HashJoinSchema::CollectFilterColumns(std::vector<FieldRef>& left_filter,
...
+Status ValidateHashJoinNodeOptions(const HashJoinNodeOptions& join_options) {
+  if (join_options.key_cmp.size() < 1 || join_options.left_keys.size() < 1 ||
+      join_options.right_keys.size() < 1) {
+    return Status::Invalid("key_cmp and keys cannot be empty");
+  }
+
+  if ((join_options.key_cmp.size() != join_options.left_keys.size()) ||
+      (join_options.key_cmp.size() != join_options.right_keys.size()) {
+    return Status::Invalid("key_cmp and keys must have the same size");
+  }
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/hash_join_node.cc:465]: In addition, the left keys and right keys must have length equal to each other.  E.g. `left_keys.size() == right_keys.size()`

**Review by westonpace — COMMENTED** [review]
I think what you have here is about as much as we can do without knowing the schemas of the input tables (which we aren't guaranteed to know at this point).

Once the schemas are bound there is actually a lot more validation we can do.  The data types of the left/right keys should match.  Only non-nested data types are allowed to be key columns.  We need to make sure all the field refs are valid.  The output of the expression should be boolean.

ARROW-15498 will add a `PrepareToProduce` function to `ExecNode` which is run after the plan is created (and thus the schemas are fixed) and before we start producing.  Maybe we can create a follow-up JIRA for the additional validation.

File: cpp/src/arrow/compute/exec/hash_join_node.cc [hunk 2]
```diff
@@ -453,6 +453,20 @@ Status HashJoinSchema::CollectFilterColumns(std::vector<FieldRef>& left_filter,
   return Status::OK();
 }
 
+Status ValidateHashJoinNodeOptions(const HashJoinNodeOptions& join_options) {
+  if (join_options.key_cmp.size() < 1 || join_options.left_keys.size() < 1 ||
+      join_options.right_keys.size() < 1) {
```
Comment (vibhatha) [inline cpp/src/arrow/compute/exec/hash_join_node.cc:458]: Nit: Wouldn't it be better to call `empty()` on vectors?

For instance:

```c++
if (join_options.empty()  || join_options.left_keys.empty() < 1 ||  join_options.right_keys.empty())
```

Comment (vibhatha) [reply cpp/src/arrow/compute/exec/hash_join_node.cc:458]: It matches the Status message as well.

File: cpp/src/arrow/compute/exec/hash_join_node.cc (hunk 1 above)
Comment (JabariBooker) [reply cpp/src/arrow/compute/exec/hash_join_node.cc:465]: This is implied in the logic of this if statement. If `key_cmp.size() == left_keys.size()` and `key_cmp.size() == right_keys.size()`, then `left_keys.size() == right_keys.size()` must be true as well.

**Review by westonpace — APPROVED** [review]
Thanks for adding this.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Hmm, the docs CI is persistently failing. Do you mind rebasing @JabariBooker?

**Comment by JabariBooker** [comment]
> Hmm, the docs CI is persistently failing. Do you mind rebasing @JabariBooker?

@lidavidm Seems to have fixed the issue. Now there are failures for `TestS3FS` and `TestS3FSGeneric` due to an IOError.

**Comment by lidavidm** [comment]
Thanks!

**Pull request merged**